src/Components/Endpoints/src/Rendering/SSRRenderModeBoundary.cs

```diff
@@ -92,11 +93,12 @@ private void Prerender(RenderTreeBuilder builder)
         builder.CloseComponent();
     }
 
-    public (ServerComponentMarker?, WebAssemblyComponentMarker?) ToMarkers(HttpContext httpContext)
+    public (ServerComponentMarker?, WebAssemblyComponentMarker?) ToMarkers(HttpContext httpContext, int sequence, object? key)
     {
         var parameters = _latestParameters is null
             ? ParameterView.Empty
             : ParameterView.FromDictionary((IDictionary<string, object?>)_latestParameters);
+        var markerKey = GenerateMarkerKey(sequence, key);
```

Comment:
Is the idea here that `@key` is "best effort"? Meaning:
1) It's not required.
2) It might not produce a unique value.
3) It might not produce a stable version between renders.

Or do we need it to produce a stable identifier for it to work. If not, what is the outcome?

Reply:
> It's not required.

Correct. If it's not supplied, the sequence number and component type will be used to distinguish between component "instances".

> It might not produce a unique value... It might not produce a stable version between renders.

Right, if a `@key` is provided, it needs to produce a stable identifier. Maybe we could require that they `@key` implement `IFormattable`, and we use the formatted output in the key we generate for the marker. And if the `@key` happens to not implement `IFormattable`, we fall back on the hash code and log a warning. Worst case, the unstable key causes the component to get reinstantiated, but it shouldn't break app functionality in general. Or, alternatively, we could throw an exception to minimize the risk of shipping an app with this mistake.

Reply:
It's more than just the `@key` being best-effort. The *entire* matching and instance retention mechanism is best-effort. There can be both:

 * False negatives, where we don't retain an interactive instance because of a `@key` not being stable across renders/servers
 * False positives, where we do retain an interactive instance and supply parameters that the developer might think would be for a new instance, because of there being no `@key` or it clashing

In both cases this has to be *not* a correctness or security issue. How I think we should think of it:

* Each time there's an SSR render, the SSR server is signing some tokens that each give permission to have an interactive component of a particular type with a particular set of parameters.
* The client maintains a *separate* relationship with an interactive component server (might not be the same as the SSR server). Within this circuit, there are a set of interactive root components, and has a ComponentId for each of them. Whenever the client receives a token from an SSR server, the client is free choose either to:
  * Start a new interactive root component using this token, getting back a new ComponentId
  * Or, update *any* existing ComponentId in the circuit to use the parameters from this token. The only restriction is that the token encodes the component type, and the server will only accept the update if the ComponentId's existing type matches what the token says.
... and each time a token is sent to the interactive server, that token is burned and cannot be replayed.

Besides that, there are no rules. Neither the SSR server nor the interactive server can say whether any given match is expected/correct or not. Only the client maintains the state to decide whether to match or not, and an untrusted client can supply any token to update any ComponentId. But that's not a problem because by definition it's allowed for any root component to receive new parameters as long as they are from one of these signed tokens (and the component type matches).

The only real problem we're trying to avoid is accidentally getting exceptions from well-behaved clients getting false-positive matches that are rejected by the interactive server.

Does this match your expectations?